### DISCUSSION TITLE
span_ruler is not working on ENT patterns

### DISCUSSION THREAD

**Discussion by kamlesh0606** [explosion/spaCy#12978] [Help: Other Questions] [answered]
Text : "A 50-year-old male patient lodges chief complaint pain in lower right side of chest for three days associated with burning in epigastrium. History is cuffed with sputum."

Get Ent in Doc :
0 | 50-year | 2 | 5 | Age
0 | male | 7 | 8 | Sex
0 | pain in lower right side of chest | 12 | 19 | chief complaint
0 | three days | 20 | 22 | time_unit
0 | burning in epigastrium | 24 | 27 | chief complaint
![image](https://github.com/explosion/spaCy/assets/3240948/8a3d1f2f-b6b7-4fbf-8621-db68845e3874)

Now , 

Add span_ruler in Pipe.
`patterns = srsly.read_jsonl("./drugs_patterns_3.jsonl")
span_ruler = nlp.add_pipe("span_ruler" , name= "span_ruler_3" , after="entity_ruler_2" , config=  { } )
span_ruler.add_patterns(patterns)`

drugs_patterns_3.jsonl File have 1 patterns ,
`{"label":"cc_time_unit","pattern":[ {"ENT_TYPE": "cc"} , {"LOWER": "for"} , {"ENT_TYPE": "time_unit"} ]}`

Output came : 
![image](https://github.com/explosion/spaCy/assets/3240948/1a322542-cfa0-4e2e-8c50-d11cb2844744)

But Output come as per logic ,
0 | pain in lower right side of chest | 12 | 19 | chief complaint
0 | three days | 20 | 22 | time_unit

## How to reproduce the behaviour

Above Side Explain All Details. Still if any need , Please update to me.

## Your Environment

* Operating System: Windows 11 
* Python Version Used: 3.7
* spaCy Version Used: 3.6
* Environment Information:

**Comment by adrianeboyd** [comment, marked as answer]
Using token patterns, each `{}` pattern matches exactly one token by default, so for this pattern you end up with the matched token "chest":

```
{"ENT_TYPE": "cc"}
```

To match one or more tokens, you can use:

```
{"ENT_TYPE": "cc", "OP": "+"}
```

Docs on operators: https://spacy.io/usage/rule-based-matching#quantifiers

If you might ever have two adjacent entity spans (e.g. something like "following a week of [intense headaches] [chest pain] for three days") and you only want to match one of the spans, you'd need to also include the `ENT_IOB` features so you can match only one entity based on the IOB tags:

```
{"ENT_IOB": "B", "ENT_TYPE": "cc"}, {"ENT_IOB": "I", "ENT_TYPE": "cc", "OP": "?"}
```

Alternatively, you could merge all entities into a single token before running patterns if you only need the phrase (and not the tokenization) in your results: https://spacy.io/api/pipeline-functions#merge_entities. Then `{"ENT_TYPE": "cc"}` would match "pain in lower right side of chest".

**Comment by kamlesh0606** [comment]
@adrianeboyd  Thanks For given Way.

Apply `nlp.add_pipe("merge_entities" , after="ner")` Pipe . And Merge All token of CC Ent.

Input Text : "A 50-year-old male patient lodges chief complaint pain in lower right side of chest , vomiting , high cholesterol for three days associated with burning in epigastrium for until chest pain is revealed. history of high blood pressure and high cholesterol 100 mg. "
![image](https://github.com/explosion/spaCy/assets/3240948/2efa4082-25f8-4e82-80f7-b4495ff70f2a)


Apply below pattern,
{"label":"1","pattern":[{"ENT_TYPE": "cc" } ,{ "LOWER" : "for" } , {"ENT_TYPE": "time_unit"  } ]}
![image](https://github.com/explosion/spaCy/assets/3240948/fa133594-7f27-42a8-8379-f2a98a0f363a)

{"label":"1","pattern":[{"ENT_TYPE": "cc" } , {"ENT_TYPE": "cc" } ,{ "LOWER" : "for" } , {"ENT_TYPE": "time_unit"  } ]}
![image](https://github.com/explosion/spaCy/assets/3240948/26341d50-e1f4-4ad4-a7ca-7e8a980d14d7)
And did get outcome.

**_Can Join 3 CC Ent join with Time_unit Ent ?_**

also tried 3 time pattern added then get below outcome,
{"label":"1","pattern":[{"ENT_TYPE": "cc" } ,{ "LOWER" : "for" } , {"ENT_TYPE": "time_unit"  } ]}
{"label":"1.2","pattern":[{"ENT_TYPE": "cc" } , {"TEXT" : "," , "OP" : "?"} , {"ENT_TYPE": "cc" } ,{ "LOWER" : "for" } , {"ENT_TYPE": "time_unit"  } ]}
{"label":"1.3","pattern":[{"ENT_TYPE": "cc" } , {"TEXT" : "," , "OP" : "?"} , {"ENT_TYPE": "cc" } , {"TEXT" : "," , "OP" : "?"} , {"ENT_TYPE": "cc" } ,{ "LOWER" : "for" } , {"ENT_TYPE": "time_unit"  } ]}
![image](https://github.com/explosion/spaCy/assets/3240948/0e681d27-6aa6-4d58-842b-b9a7f85cc785)
But When get 5-6 CC Token then add 6-7 time patten. Same way other place comes Same points. Please provide good option if you have. 


2nd points,
 Start with  "chief complaint" And End with ["." , "," ,"End of Statement"]. _**Please provide pattern of it if possible.**_

**Reply by adrianeboyd** [reply]
Unfortunately the `Matcher` patterns don't currently support groups, so there's not a more elegant way to write some of the these patterns.

You can make individual tokens optional with `"OP": "?"` (as you're already doing) and you can use `"IN"` to create alternatives for a single token like `"LOWER": {"IN": ["for", "until"]}`.

You might also be able to match on `"cc"`...`"time_unit"` by matching anything up to the next time phrase, but here you might want to add a maximum match length or other constraints in order to avoid long matches to some unrelated time unit that's far away in the text:

```
{"ENT_TYPE": "cc"}, {"ENT_TYPE": {"NOT_IN": ["time_unit"]}, "OP": "{,10}"}, {"ENT_TYPE": "time_unit"}
```

Once you have a list of matched spans, you can filter for the longest non-overlapping spans with [`spacy.util.filter_spans`](https://spacy.io/api/top-level#util.filter_spans), e.g. `filter_spans(doc.spans["span_ruler"])`.

**Reply by kamlesh0606** [reply]
@adrianeboyd  Thanks for given support.

Have you support groups for patterns parts ? if yes, I will take it and more discussion there.

One More Time Thanks For giving Correct Way for new Development.

**Reply by adrianeboyd** [reply]
There are some older issues related to this, but more for labeled groups than groups that can repeat with operators:

* https://github.com/explosion/spaCy/issues/4642
* https://github.com/explosion/spaCy/discussions/3275

The matcher alignments provide a kind of workaround for labeled groups, but it's still not exactly groups.

**Reply by kamlesh0606** [reply]
@adrianeboyd , Thanks for given update.

Can get Pipe By Pipe Doc Object ? Or can get process ID in side Pipe ? if yes then will manage Pipe Doc Output Obj using process ID .  

I get one way Like , "user_data" use of Doc Object. it it right way ?

**Reply by adrianeboyd** [reply]
Sorry, I'm not sure exactly what you're asking, and it sounds like it might be a separate question from this thread.

Could you open a new discussion with a code example that shows what you're trying to do?

**Reply by kamlesh0606** [reply]
ok

@adrianeboyd  , https://github.com/explosion/spaCy/discussions/12986 new Points added.
